src/python/pants/backend/java/compile/javac_binary.py

```diff
@@ -0,0 +1,69 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import logging
+import textwrap
+from dataclasses import dataclass
+from typing import ClassVar
+
+from pants.backend.java.compile.javac_subsystem import JavacSubsystem
+from pants.engine.fs import CreateDigest, Digest, FileContent, MergeDigests
+from pants.engine.rules import Get, collect_rules, rule
+from pants.jvm.resolve.coursier_setup import Coursier
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class JavacBinary:
+    digest: Digest
+    javac: ClassVar[str] = "__javac_binary/javac.sh"
+    classfiles_relpath: ClassVar[str] = "classfiles"
+
+
+@rule
+async def setup_javac_binary(coursier: Coursier, javac: JavacSubsystem) -> JavacBinary:
+    javac_wrapper_script = textwrap.dedent(
+        f"""\
+        set -eux
+
+        javac_path="$({coursier.coursier.exe} java-home --jvm {javac.options.jdk})/bin/javac"
+        /bin/mkdir -p {JavacBinary.classfiles_relpath}
+        exec "${{javac_path}}" "$@"
+        """
+    )
+
+    print("WRAPPER SCRIPT:")
```

Comment:
Leftover debugging? If you want to keep. maybe switch to `logger.debug`?

Reply:
Yeah this is deliberate, after pushing the PR I found that tests are failing in a way I can't reproduce locally.  Looks like it's probably something to do with Coursier bootstrapping the wrong architecture `javac` binary on the CI machine.  For now I need to keep debugging with a "push and wait for CI" workflow unfortunately.  I'll ping you when I've cleaned up the debugging and fixed the issue